app/demo-geant-integration/HepMC3Reader.cc

```diff
@@ -0,0 +1,162 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HepMC3Reader.cc
+//---------------------------------------------------------------------------//
+#include "HepMC3Reader.hh"
+
+#include <mutex>
+#include <G4SystemOfUnits.hh>
+#include <HepMC3/ReaderFactory.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+#include "GlobalSetup.hh"
+
+namespace
+{
+// Mutex visible to all threads
+static std::mutex mutex;
+} // namespace
+
+namespace demo_geant
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Singleton declaration.
+ */
+static HepMC3Reader* hepmc3_reader_singleton = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return non-owning pointer to a singleton.
+ */
+HepMC3Reader* HepMC3Reader::instance()
+{
+    {
+        mutex.lock();
+        if (!hepmc3_reader_singleton)
+        {
+            hepmc3_reader_singleton = new HepMC3Reader();
+        }
+        mutex.unlock();
+    }
+
+    CELER_LOG_LOCAL(status) << "Instancing HepMC3 singleton";
+    return hepmc3_reader_singleton;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add HepMC3 primaries to a Geant4 event. This function is called by
+ * `G4VUserPrimaryGeneratorAction::GeneratePrimaries`.
+ */
+void HepMC3Reader::GeneratePrimaryVertex(G4Event* g4_event)
+{
+    mutex.lock();
```

Comment:
Use `std::lock_guard` and a local static variable: see `SharedParams::locked_initialize`, `corecel/io/Logger.cc`, etc. Otherwise, forgetting to "unlock", hitting an exception, or returning early will cause the app to hang.